cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -317,18 +326,24 @@ private HttpResponse doRequest(HttpMethod method, String resource,
       return builder;
     }
     String userId;
-    if (NamespaceId.SYSTEM.equals(namespaceId)) {
-      // For getting a system dataset like MDS, use the system principal. It is ok to do so, since DatasetServiceClient
-      // is an internal client that is not exposed to users.
-      // TODO: CDAP-6583: This is dangerous. Remove.
+    if (NamespaceId.SYSTEM.equals(namespaceId) &&
+      (!kerberosEnabled || masterPrincipal.equals(UserGroupInformation.getCurrentUser().getUserName()))) {
```

Comment:
I've verified that ``UserGroupInformation.getCurrentUser().getUserName()`` always returns the full principal from master as well as master service containers.